pkg/storage/oss/oss.go

```diff
@@ -0,0 +1,662 @@
+	defer os.Remove(f.Name())
+	defer os.Unsetenv(provider.ENVCredentialFile)
+	_, err = f.Write(creds)
+	if err != nil {
+		return nil, err
+	}
+	os.Setenv(provider.ENVCredentialFile, f.Name())
+	// use Alibaba provider initialization
+	p := provider.NewProfileProvider("default")
+	// return a valid auth credential
+	return p.Resolve()
+}
+
+func (d *driver) getCredentialsConfigData() error {
+	if d.credentials != nil {
+		return nil
+	}
+
+	// Look for a user defined secret to get the Alibaba Cloud credentials from first
+	sec, err := d.Listers.Secrets.Get(defaults.ImageRegistryPrivateConfigurationUser)
+	if err != nil && errors.IsNotFound(err) {
+		// Fall back to those provided by the credential minter if nothing is provided by the user
+		sec, err = d.Listers.Secrets.Get(defaults.CloudCredentialsName)
+		if err != nil {
+			return fmt.Errorf("unable to get cluster minted credentials %q: %v", fmt.Sprintf("%s/%s", defaults.ImageRegistryOperatorNamespace, defaults.CloudCredentialsName), err)
+		}
+	}
+
+	credential, err := fetchCredentialsIniFromSecret(sec)
+	if err != nil {
+		return fmt.Errorf("failed to generate shared secrets data: %v", err)
+	}
+	// This must be an AccessKeyCrential because the oss.New requires accessKeyID and accessSecretKey
+	d.credentials = credential.(*credentials.AccessKeyCredential)
+
+	return nil
+}
+
+// isInternal return An internal endpoint or the public endpoint for OSS access.  default internal
+func (d *driver) isInternal() bool {
+	if d.Config.EndpointAccessibility == imageregistryv1.PublicEndpoint {
+		return false
+	}
+	return true
+}
+
+// getOSSEndpoint returns an endpoint that allows us to interact
+// with the Alibaba Cloud OSS service, details in https://www.alibabacloud.com/help/doc-detail/31837.htm
+func (d *driver) getOSSEndpoint() string {
+	if d.isInternal() {
+		return fmt.Sprintf("https://oss-%s-internal.aliyuncs.com", d.Config.Region)
+	}
+	return fmt.Sprintf("https://oss-%s.aliyuncs.com", d.Config.Region)
+}
+
+func (d *driver) getOSSService() (*oss.Client, error) {
+	err := d.getCredentialsConfigData()
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.UpdateEffectiveConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint := d.getOSSEndpoint()
+
+	var clientOptions oss.ClientOption
+	if d.roundTripper != nil {
+		clientOptions = oss.HTTPClient(&http.Client{
+			Transport: d.roundTripper,
+		})
+	}
+	client, err := oss.New(endpoint, d.credentials.AccessKeyId, d.credentials.AccessKeySecret, clientOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	return client, err
+}
+
+// ConfigEnv configures the environment variables
+// Note: it is the callers responsiblity to make sure the returned file
+// location is cleaned up after it is no longer needed.
+func (d *driver) ConfigEnv() (envs envvar.List, err error) {
+	err = d.UpdateEffectiveConfig()
+	if err != nil {
+		return
+	}
+
+	envs = append(envs,
+		envvar.EnvVar{Name: "REGISTRY_STORAGE_OSS_REGIONENDPOINT", Value: d.getOSSEndpoint()},
```

Comment:
I have not tested `REGISTRY_STORAGE_OSS_REGIONENDPOINT`. 

I had tested
```
envs = append(envs, envvar.EnvVar{Name: envRegistryStorageOssEndpoint, Value: d.getOSSEndpoint})
```
This did not work. When passing this into the image registry code it would break the image-registry it would result in this:
```
2021/12/13 23:52:10 GET https://oss-us-east-1.aliyuncs.com/?delimiter=&marker=&max-keys=1&prefix= ...
panic: Aliyun API Error: RequestId: 61B7DCAACD80323433AF847C Status Code: 403 Code: SignatureDoesNotMatch Message: The request signature we calculated does not match the signature you provided. Check your key and signing method.
```
Notice the URL above. The URL is missing the bucket name. I fixed this by doing either of 2 things. 
1. I removed this variable and it worked. 
2. I modified the `OSS_ENDPOINT` to look like this and it worked:
```
	envs = append(envs, envvar.EnvVar{Name: envRegistryStorageOssEndpoint, Value: fmt.Sprintf("%s.%s", d.Config.Bucket, d.getOSSEndpoint())})
```

I see that you have modified the variable to `REGISTRY_STORAGE_OSS_REGIONENDPOINT`. Is this correct? I haven't tested with this one yet but not sure if this is valid or not.

Reply:
envRegistryStorageOssEndpoint and "REGISTRY_STORAGE_OSS_REGIONENDPOINT"  are the same.

Reply:
@menglingwei I can't see this parameter (`regionendpoint`) from the [OSS docker storage documentation](https://docs.docker.com/registry/storage-drivers/oss/). Does it need to be compatible?

Reply:
> @menglingwei I can't see this parameter (`regionendpoint`) from the [OSS docker storage documentation](https://docs.docker.com/registry/storage-drivers/oss/). Does it need to be compatible?

@mtulio  change to REGISTRY_STORAGE_OSS_ENDPOINT